[improve][broker] PIP-192 Added ServiceUnitStateChannel interface

heesung-sn opened the pull request:
<!--
### Contribution Checklist
  
  - PR title format should be *[type][component] summary*. For details, see *[Guideline - Pulsar PR Naming Convention](https://docs.google.com/document/d/1d8Pw6ZbWk-_pCKdOmdvx9rnhPiyuxwq60_TrD68d7BA/edit#heading=h.trs9rsex3xom)*. 

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.
-->

<!-- Either this PR fixes an issue, -->

Fixes #xyz

<!-- or this PR is one task of an issue -->

Master Issue: https://github.com/apache/pulsar/issues/16691

### Motivation

<!-- Explain here the context, and why you're making that change. What is the problem you're trying to solve. -->
We will start raising PRs to implement PIP-192, https://github.com/apache/pulsar/issues/16691

### Modifications
The PR adds base classes for the new broker load balance project and does not integrate with the existing load balance logic --this PR should not impact the existing broker load balance behavior. 

For the pip-192 project, this PR
- defines the classes under `org.apache.pulsar.broker.loadbalance.extensions` package.
- defines this ServiceUnitStateChannel public interface and its expected behaviors.
- defines ServiceUnitState and ServiceUnitStateData data classes.
- defines Split and Unload data classes.




<!-- Describe the modifications you've done. -->

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

  - *Added unit tests for the new classes.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] Anything that affects deployment

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. Please attach the local preview screenshots (run `sh start.sh` at `pulsar/site2/website`) to your PR description, or else your PR might not get merged. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

We will have separate PRs to update the Doc later.

### Matching PR in forked repository

PR in forked repository: https://github.com/heesung-sn/pulsar/pull/11

<!--
After opening this PR, the build in apache/pulsar will fail and instructions will
be provided for opening a PR in the PR author's forked repository.

apache/pulsar pull requests should be first tested in your own fork since the 
apache/pulsar CI based on GitHub Actions has constrained resources and quota.
GitHub Actions provides separate quota for pull requests that are executed in 
a forked repository.

The tests will be run in the forked repository until all PR review comments have
been handled, the tests pass and the PR is approved by a reviewer.
-->

tisonkun commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensible/channel/ServiceUnitStateData.java:
```diff
@@ -0,0 +1,50 @@
+ * Defines data for the service unit state changes.
+ * This data will be broadcast in ServiceUnitStateChannel.
+ */
+@Data
+@ToString
+@NoArgsConstructor
+@AllArgsConstructor
+public class ServiceUnitStateData {
+    private ServiceUnitState state;
+    private String broker;
+    private String sourceBroker;
+    private long timestamp;
```
1. `@Data` already infers `@ToString`.
2. Since the broker JDK version is upgraded to 17, do you think of using [record class](https://docs.oracle.com/en/java/javase/17/language/records.html) to define data models?
3. Are these field intended to be mutable?

heesung-sn replied:
Updated to `record` classes.

tisonkun commented on pulsar-broker/src/test/java/org/apache/pulsar/broker/loadbalance/extensible/channel/ServiceUnitStateDataTest.java:
```diff
@@ -0,0 +1,46 @@
+        assertEquals(data1.getState(), Owned);
+        assertEquals(data1.getBroker(), "A");
+        assertNull(data1.getSourceBroker());
+        assertTrue(data1.getTimestamp() > 0);
```
You can use `assertj`'s assertion for better error message:

```java
import static org.assertj.core.api.Assertions.assertThat;

assertThat(data1.getTimestamp()).isGreaterThan(0);
```

tisonkun replied:
ditto other assertions.

heesung-sn replied:
Updated.

tisonkun commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensible/channel/ServiceUnitState.java:
```diff
@@ -0,0 +1,99 @@
+
+    Splitting; // the service unit(e.g. bundle) is in the process of splitting.
+    private static Map<ServiceUnitState, Set<ServiceUnitState>> validTransitions = new HashMap<>() {{
+        put(null, new HashSet<>() {{
```
Shall we define a state `Free`/`NotOwned` over `null`? Then we can avoid handling business `null`s.

heesung-sn replied:
Updated.

tisonkun commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensible/channel/ServiceUnitState.java:
```diff
@@ -0,0 +1,99 @@
+
+    Splitting; // the service unit(e.g. bundle) is in the process of splitting.
+    private static Map<ServiceUnitState, Set<ServiceUnitState>> validTransitions = new HashMap<>() {{
+        put(null, new HashSet<>() {{
```
I'd prefer immutable sets over `HashSet` here. For example, `Set.of(...)`.

heesung-sn replied:
Updated.

tisonkun reviewed (commented):
Thanks for your contribution @heesung-sn! Comments inline.

Since we turn on the semantic PR title checks, it can edit the PR title as: `[improve][broker] Added ServiceUnitStateChannel interface`

tisonkun reviewed (approved):
LGTM. Thanks for your contribution!

cc @codelipenghui @Jason918 @Demogorgon314 @eolivelli

tisonkun commented:
/pulsarbot run-failure-checks

Technoboy- reviewed (approved):
LGTM

heesung-sn commented:
- Updated the package name to `extensions` according to the comment in https://github.com/apache/pulsar/pull/18084#issuecomment-1309449780.

- added the function close() in ServiceUnitStateChannel.

- minor function name updates.

Jason918 commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitState.java:
```diff
@@ -0,0 +1,81 @@
+ *    │           │   discard(parent)│          │
+ *    └───────────┘                  └──────────┘
+ */
+public enum ServiceUnitState {
```
Is this `ServiceUnit` the same as `bundle` for now and in the foreseeable future?

eolivelli replied:
I would like to keep open the door to having also a single topic (partition) and not only a "bundle"

heesung-sn replied:
I see two conflicting comments here.

The intention is to keep this interface open for other service units(topic, etc). But the initial target service units will be bundles only.

Jason918 replied:
> The intention is to keep this interface open for other service units(topic, etc). But the initial target service units will be bundles only.

Does support for other "service units" included in PIP-192?

Both is fine for me, as long as we don't mixing them up. If we stick to service unit, so there won't be any 'bundle' appear in load balance module unless it's some specific implementation, right?

heesung-sn replied:
For pip-192, no, it will be bundle only. 

Sure, we can stick to "serviceUnit" for var and func names.

Jason918 commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/models/Split.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Defines the information required for a bundle split.
+ */
+public record Split(String bundle, String sourceBroker, Map<String, Optional<String>> splitBundleToDestBroker) {
```
It's confusing that we use "bundle" here, but use "serviceUnit" in `Unload`.

Jason918 reviewed (commented):
Overall LGTM.
My only concern is that can we just use "bundle" instead of introducing the concept "ServiceUnit". It's confusing and a bit of over design to me. 
I'm aware that we already have some "service unit" in current codebase. It's already confusing when I first saw that, so can we unify or simplify the concept in this PIP.

heesung-sn commented:
@Jason918  @eolivelli,
I updated the code based on your comments. Plz, take a look again.

heesung-sn commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/models/Unload.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Defines the information required to unload or transfer a service unit(e.g. bundle).
+ */
+public record Unload(String sourceBroker, String serviceUnit, Optional<String> destBroker) {
```
I don't see Record related issues right now, and we could easily switch back to the lombok + pojo if blocked by any. By default, I think it is a good practice to try Record as much as we can.

eolivelli reviewed (approved):
+1

heesung-sn commented:
@Jason918 can you take a look at again by any chance? We need to merge this first to raise next PRs.

Jason918 commented:
> @Jason918 can you take a look again by any chance? We need to merge this first to raise next PRs.

Sorry for the late reply. This PR LGTM. :)
